### PR TITLE
process block and balance transfer transaction

### PR DESCRIPTION
**Pull request by dorisRiver**

### CODE REVIEW

File: cmd/indexer/main.go [hunk 1]
```diff
@@ -60,7 +63,16 @@ func exec(ctx *cli.Context) error {
 		log.Fatal("initialize mysql db error", "error", err)
 	}
 
-	indexer.NewEventProcessor(cfg.RefreshInterval, cfg.NodeEndpoint, db).Run(ctx.Context)
+	eventProcessor := indexer.NewEventProcessor(ctx.Context, cfg.RefreshInterval, cfg.NodeEndpoint, db)
+	go eventProcessor.Run()
+	c := make(chan os.Signal, 1)
```
Comment (maxkaneco) [inline cmd/indexer/main.go:68]: ch for channel

File: cmd/indexer/main.go [hunk 2]
```diff
@@ -60,7 +63,16 @@ func exec(ctx *cli.Context) error {
...
+	signal.Notify(c, syscall.SIGQUIT, syscall.SIGTERM, syscall.SIGINT)
+	for sig := range c {
```
Comment (maxkaneco) [inline cmd/indexer/main.go:70]: let's copy the implementation from  chain/node/node.go?

File: database/orm/transaction.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		"BALANCE_TRANSFER":  BalanceTransfer,
+		"VALIDATOR_DEPOSIT": ValidatorDeposit,
+		"VALIDATOR_EXIT":    ValidatorExit,
+		"AUDITOR_DEPOSIT":   AuditorDeposit,
+		"AUDITOR_EXIT":      AuditorExit,
+		"OBJECT_COMMIT":     ObjectCommit,
+		"OBJECT_AUDIT":      ObjectAudit,
+		"OBJECT_POR":        ObjectPor,
+	}
+)
```
Comment (maxkaneco) [inline database/orm/transaction.go:44]: use types and names from proto?

File: indexer/block.go [hunk 4]
```diff
@@ -1,14 +1,134 @@
...
+		return err
+	}
+
 	e.currentSlot = block.Slot
 	e.currentHash = block.BlockHash
 	return nil
 }
 
+func processAttestations(db *gorm.DB, blockID uint64, attestations []*gateway.Attestation) error {
+	as := make([]*orm.Attestation, len(attestations))
```
Comment (maxkaneco) [inline indexer/block.go:48]: atts

File: indexer/block.go [hunk 5]
```diff
@@ -1,14 +1,134 @@
...
+	for i, a := range attestations {
+		bits := strings.Replace(
```
Comment (maxkaneco) [inline indexer/block.go:50]: Strings.Replace is a bit hacky.
https://stackoverflow.com/questions/37532255/one-liner-to-transform-int-into-string

File: indexer/block.go [hunk 6]
```diff
@@ -1,14 +1,134 @@
...
+	}); err != nil {
+		return err
+	}
+
 	e.currentSlot = block.Slot
 	e.currentHash = block.BlockHash
 	return nil
 }
 
+func processAttestations(db *gorm.DB, blockID uint64, attestations []*gateway.Attestation) error {
```
Comment (maxkaneco) [inline indexer/block.go:47]: too long

File: indexer/block.go [hunk 7]
```diff
@@ -1,14 +1,134 @@
...
+			SourceHash:      a.Source.Hash,
+			TargetEpoch:     a.Target.Epoch,
+			TargetHash:      a.Target.Hash,
+			Signature:       a.Signature,
+		}
+	}
+	return db.Create(as).Error
+}
+
+func (e *EventProcessor) processTransactions(db *gorm.DB, blockID uint64, transactions []*gateway.Tx) error {
```
Comment (maxkaneco) [inline indexer/block.go:70]: too long

File: indexer/block.go [hunk 8]
```diff
@@ -1,14 +1,134 @@
...
+
+			if err := e.applyTransfer(db, t.BalanceTransfer.To, int64(t.BalanceTransfer.Amount)); err != nil {
+				return err
+			}
+		}
+	}
+	return db.Create(txs).Error
+}
+
+func (e *EventProcessor) applyTransfer(db *gorm.DB, address string, changeAmount int64) error {
```
Comment (maxkaneco) [inline indexer/block.go:101]: too long.
This func updates account balance, not "apply transfer"

File: indexer/block.go [hunk 9]
```diff
@@ -1,14 +1,134 @@
...
+			Type:     orm.StrToType(t.Type),
+			Raw:      string(raw),
+		}
+
+		if t.Type == orm.BalanceTransfer.String() {
+			if err := e.applyTransfer(db, t.From, -1*int64(t.BalanceTransfer.Amount)); err != nil {
+				return err
+			}
+
+			if err := e.applyTransfer(db, t.BalanceTransfer.To, int64(t.BalanceTransfer.Amount)); err != nil {
```
Comment (maxkaneco) [inline indexer/block.go:93]: format

File: indexer/eventprocessor.go [hunk 10]
```diff
@@ -2,69 +2,100 @@ package indexer
...
+
 // Run executing the timing task of processing chain data.
-func (e *EventProcessor) Run(ctx context.Context) {
+func (e *EventProcessor) Run() {
 	ticker := time.NewTicker(time.Duration(e.refreshInterval) * time.Second)
 	defer ticker.Stop()
 
-	for {
+	e.wg.Add(1)
+	for ; true; <-ticker.C {
```
Comment (maxkaneco) [inline indexer/eventprocessor.go:71]: this implementation is not ideal. When it waits on ticker.C, it won't process e.quit and e.ctx.Done() signal

File: chain/node.go [hunk 11]
```diff
@@ -52,6 +53,13 @@ func (n *NodeClient) BlockByHash(ctx context.Context, hash string) (*gateway.Blo
 	return b, httpGet(ctx, url, b)
 }
 
+// Account gets account detail by account address
+func (n *NodeClient) Account(ctx context.Context, address string) (*gateway.AccountResp, error) {
```
Comment (maxkaneco) [inline chain/node.go:57]: shall we perform a sanity check for the address before sending out req? This applies to other APIs

File: database/orm/transaction.go (hunk 3 above)
Comment (dorisRiver) [reply database/orm/transaction.go:44]: if the proto changes, the type of all transactions in the database may be  changed

File: indexer/block.go (hunk 8 above)
Comment (dorisRiver) [reply indexer/block.go:101]: It not only modifies the balance but also modifies the nonce, updateAccount for the func name?

File: database/orm/transaction.go (hunk 3 above)
Comment (maxkaneco) [reply database/orm/transaction.go:44]: Proto cannot change after launch. It is part of consensus and it in State db

File: chain/node.go [hunk 12]
```diff
@@ -52,6 +59,43 @@ func (n *NodeClient) BlockByHash(ctx context.Context, hash string) (*gateway.Blo
...
+	if !isValidPublicKey(pk) {
+		return nil, errPublicKey
+	}
+
+	url := fmt.Sprintf("%s/%s?public_key=%s", n.endpoint, account, pk)
+	a := &gateway.AccountResp{}
+	return a, httpGet(ctx, url, a)
+}
+
+func isValidPublicKey(pk string) bool {
```
Comment (maxkaneco) [inline chain/node.go:73]: this is over complicated.
1. strings.ToLower
2. strings.HasPrefix("0x")
3. hex.DecodeString

File: indexer/block.go [hunk 13]
```diff
@@ -1,14 +1,153 @@
...
+				int64(t.BalanceTransfer.Amount),
+			); err != nil {
+				return err
+			}
+		}
+	}
+	return db.Create(txs).Error
+}
+
+func (e *EventProcessor) applyTransfer(
```
Comment (maxkaneco) [inline indexer/block.go:116]: again, this function is not performing transfer. It only updates a single account.

File: indexer/block.go [hunk 14]
```diff
@@ -1,14 +1,153 @@
...
+			Hash:     t.TxHash,
+			From:     t.From,
+			Position: uint64(i),
+			GasPrice: t.GasPrice,
+			Type:     pbc.TxType_value[t.Type],
+			Raw:      string(raw),
+		}
+
+		if t.Type == pbc.TxType_BALANCE_TRANSFER.String() {
+			if err := e.applyTransfer(
```
Comment (maxkaneco) [inline indexer/block.go:96]: shall we make these two updates in one sql transaction?

File: indexer/eventprocessor.go [hunk 15]
```diff
@@ -2,50 +2,76 @@ package indexer
...
 		node:            chain.NewNodeClient(nodeEndpoint),
 		currentSlot:     currentSlot,
 		currentHash:     currentHash,
+		quit:            make(chan struct{}),
 	}
 }
 
+// Stop exit event processor
+func (e *EventProcessor) Stop() {
+	if err := e.db.Model(&orm.ChainStatus{}).Where("id = 1").Updates(
```
Comment (maxkaneco) [inline indexer/eventprocessor.go:54]: What would happen if this update is not properly done? e.g. indexer crashed

File: indexer/eventprocessor.go [hunk 16]
```diff
@@ -57,13 +83,17 @@ func (e *EventProcessor) Run(ctx context.Context) {
 			}
 
 			for e.currentSlot < headSlot {
-				if err := e.processEvents(ctx); err != nil {
+				if err := e.processEvents(e.ctx); err != nil {
 					log.Error("keeper fail on sync chain events", "error", err)
```
Comment (maxkaneco) [inline indexer/eventprocessor.go:87]: keeper?

File: indexer/eventprocessor.go [hunk 17]
```diff
@@ -2,50 +2,76 @@ package indexer
...
+func (e *EventProcessor) Stop() {
+	if err := e.db.Model(&orm.ChainStatus{}).Where("id = 1").Updates(
+		map[string]interface{}{
+			"slot": e.currentSlot,
+			"hash": e.currentHash,
+		}).Error; err != nil {
+		log.Fatal("stop updating chain status failed", err)
+	}
+	close(e.quit)
+	e.wg.Wait()
```
Comment (maxkaneco) [inline indexer/eventprocessor.go:62]: Is this wait correct? Other parts of the indexer need to wait for Stop() to complete, not these opposite?

File: indexer/block.go [hunk 18]
```diff
@@ -1,14 +1,153 @@
...
+		}
+
+		txs[i] = &orm.Transaction{
+			BlockID:  blockID,
+			Hash:     t.TxHash,
+			From:     t.From,
+			Position: uint64(i),
+			GasPrice: t.GasPrice,
+			Type:     pbc.TxType_value[t.Type],
+			Raw:      string(raw),
```
Comment (maxkaneco) [inline indexer/block.go:92]: why do we cast to string type? Doesn't ORM support binary/[]byte type field?

Comment (maxkaneco) [inline indexer/block.go:92]: why do we cast to string type? Doesn't ORM support binary/[]byte type field?

File: chain/node.go [hunk 19]
```diff
@@ -52,6 +61,25 @@ func (n *NodeClient) BlockByHash(ctx context.Context, hash string) (*gateway.Blo
...
+	if strings.HasPrefix("0x", strings.ToLower(pk)) {
+		pk = pk[2:]
+	}
+	_, err := hex.DecodeString(pk)
+	return len(pk) == bls.PublicKeyLength*2 && err == nil
```
Comment (maxkaneco) [inline chain/node.go:80]: this check does not guarantee pk validity

File: indexer/eventprocessor.go [hunk 20]
```diff
@@ -7,78 +7,111 @@ import (
...
+		headSlot, err := e.node.HeadSlot(e.ctx)
+		if err != nil {
+			log.Error("request head slot from photon node failed", "error", err)
+			continue
+		}
 
-		case <-ctx.Done():
-			return
+		if e.currentSlot >= headSlot {
+			log.Info("local slot is best slot")
```
Comment (maxkaneco) [inline indexer/eventprocessor.go:74]: this msg will be spammy

File: indexer/eventprocessor.go [hunk 21]
```diff
@@ -7,78 +7,111 @@ import (
...
 
 // NewEventProcessor returns the new instance of EventProcessor.
-func NewEventProcessor(refreshInterval uint64, nodeEndpoint string, db *gorm.DB) *EventProcessor {
+func NewEventProcessor(
+	ctx context.Context,
+	refreshInterval uint64,
+	nodeEndpoint string,
+	db *gorm.DB,
+) *EventProcessor {
 	currentSlot, currentHash, err := currentChainStatus(db)
```
Comment (maxkaneco) [inline indexer/eventprocessor.go:34]: It is better to move this fetch to Run() and retry until it succeeds. curentSlot and currentHash can be local var.

**Review by maxkaneco — APPROVED** [review]

**Comment by maxkaneco** [comment]
PPH=4hrs

**Pull request merged**